### PR TITLE
cannon: support basic array functionality (length(), get, set)

### PR DESCRIPTION
**Pull request by hzahradnik**
As discussed, and mentioned in the previous PR, I am currently working to support the basic array methods: 
* get 
* set
* length()
For now, I finished the implementation for `length()`, which has its own bytecode `ArrayLength` with two operands: destination register of the result and the register with the array reference. Cannon takes these operands and checks for nil before loading the length.

Additionally, I removed some unused Intrinsics and added a explicit bytecode for an array boundary check.

### CODE REVIEW

File: dora/src/bytecode/dumper.rs [hunk 1]
```diff
@@ -920,6 +920,13 @@ impl<'a> BytecodeVisitor for BytecodeDumper<'a> {
         self.emit_new_array("NewArray", dest, cls, length);
     }
 
+    fn visit_array_length(&mut self, dest: Register, arr: Register) {
+        self.emit_reg2("ArrayLength", dest, arr);
+    }
+    fn visit_array_bound_check(&mut self, arr: Register) {
```
Comment (dinfuehr) [inline dora/src/bytecode/dumper.rs]: I think this needs a second register for the index.

Comment (hzahradnik) [reply dora/src/bytecode/dumper.rs]: Of course, don't know how i missed that.

File: dora/src/bytecode/data.rs [hunk 2]
```diff
@@ -380,6 +401,8 @@ impl BytecodeOpcode {
             | BytecodeOpcode::InvokeStaticPtr
             | BytecodeOpcode::NewObject
             | BytecodeOpcode::NewArray
```
Comment (dinfuehr) [inline dora/src/bytecode/data.rs]: All the LoadArrayXXX and StoreArrayXXX also need the position

File: dora/src/bytecode/generator.rs [hunk 3]
```diff
@@ -1001,6 +1008,21 @@ impl<'a, 'ast> AstBytecodeGen<'a, 'ast> {
         }
     }
 
+    fn emit_intrinsic_array_len(&mut self, arr: &Expr, pos: Position, dest: DataDest) -> Register {
+        if dest.is_effect() {
+            self.visit_expr(arr, dest);
```
Comment (dinfuehr) [inline dora/src/bytecode/generator.rs]: I don't think this is correct, e.g. `nil.length();` should still throw but AFAIU it doesn't

File: dora/src/bytecode/reader.rs [hunk 4]
```diff
@@ -1117,6 +1117,115 @@ where
...
+            }
+            BytecodeOpcode::LoadArrayPtr => {
+                let src = self.read_register(wide);
+                let array = self.read_register(wide);
+                let index = self.read_register(wide);
+                self.visitor.visit_load_array_ptr(src, array, index);
+            }
+
+            BytecodeOpcode::StoreArrayBool => {
+                let src = self.read_register(wide);
```
Comment (dinfuehr) [inline dora/src/bytecode/reader.rs]: dest instead of src

File: dora/src/bytecode/reader.rs [hunk 5]
```diff
@@ -1117,6 +1117,115 @@ where
...
+                self.visitor.visit_array_length(dest, array);
+            }
+            BytecodeOpcode::ArrayBoundCheck => {
+                let array = self.read_register(wide);
+                let index = self.read_register(wide);
+                self.visitor.visit_array_bound_check(array, index);
+            }
+
+            BytecodeOpcode::LoadArrayBool => {
+                let src = self.read_register(wide);
```
Comment (dinfuehr) [inline dora/src/bytecode/reader.rs]: dest instead of src

Comment (hzahradnik) [reply dora/src/bytecode/reader.rs]: Done.

File: dora/src/bytecode/data.rs (hunk 2 above)
Comment (hzahradnik) [reply dora/src/bytecode/data.rs]: Done.

File: dora/src/bytecode/generator.rs (hunk 3 above)
Comment (hzahradnik) [reply dora/src/bytecode/generator.rs]: As we talked about a explicit `NilCheck`: Wouldn't it a possibility to use this here? We aren't interested in the result, but only want to check, if the object is nil.

**Comment by hzahradnik** [comment]
I've more or less completed the implementation of arrays, but some things are still missing.
What is now supported in cannon:

- allocation of a new array with `let array = Array[T](8*8)`
- retrieving of allocated size of an array with `let length = array.length()`. If this expression is an effect, it will only check, if the left handside is nil. (see `NilCheck`)
- accessing of a value with `let value = array(length-1)`
- setting value with `array(length-1) = value-1`

What is missing:

- Unit `()` and Tuples
- `ArrayBoundCheck` is not implemented. Is needed when accessing a Unit array element. (Maybe also needed when array access is effect, as we don't need the value, but have to check for nil and array bounds.)

File: dora/src/bytecode/generator.rs [hunk 6]
```diff
@@ -1375,7 +1439,58 @@ impl<'a, 'ast> AstBytecodeGen<'a, 'ast> {
                         }
                     }
                 }
-                _ => unreachable!(),
+                ExprCall(ref _call) => {
+                    if let Some(intrinsic) = self.get_intrinsic(expr.id) {
+                        match intrinsic {
+                            Intrinsic::GenericArraySet => {
+                                let object = &_call.callee;
```
Comment (dinfuehr) [inline dora/src/bytecode/generator.rs]: Can you rename `_call` to `call`

Comment (hzahradnik) [reply dora/src/bytecode/generator.rs]: Done

**Comment by dinfuehr** [comment]
Cool, thanks!

**Pull request merged**